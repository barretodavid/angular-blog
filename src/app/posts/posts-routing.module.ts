import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';

import { PostCreateEditComponent } from './components/post-create-edit/post-create-edit.component';
import { PostListComponent } from './components/post-list/post-list.component';

const routes: Routes = [
  { path: '', component: PostListComponent },
  { path: 'create-post', component: PostCreateEditComponent },
  { path: 'edit-post/:uuid', component: PostCreateEditComponent },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class PostsRoutingModule { }
