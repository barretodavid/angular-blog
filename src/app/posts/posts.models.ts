export type Post = Readonly<{
  uuid: string;
  title: string;
  content: string;
}>;
