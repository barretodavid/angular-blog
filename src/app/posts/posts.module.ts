import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';

import { MatCardModule, MatButtonModule } from '@angular/material';

import { PostsRoutingModule } from './posts-routing.module';
import { PostCreateEditComponent } from './components/post-create-edit/post-create-edit.component';
import { PostListComponent } from './components/post-list/post-list.component';
import { PostSummaryComponent } from './components/post-summary/post-summary.component';

import { SharedModule } from 'src/app/shared/shared.module';
import { StoreModule } from '@ngrx/store';
import { postsReducer } from './store/posts.reducer';
import { EffectsModule } from '@ngrx/effects';
import { PostEffects } from './store/posts.effects';

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MatCardModule,
    MatButtonModule,
    StoreModule.forFeature('posts', postsReducer),
    EffectsModule.forFeature([PostEffects]),
    PostsRoutingModule,
    SharedModule,
  ],
  declarations: [
    PostCreateEditComponent,
    PostListComponent,
    PostSummaryComponent,
  ],
  exports: [],
})
export class PostsModule {}
