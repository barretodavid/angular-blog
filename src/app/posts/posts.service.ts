import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';

import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { Post } from './posts.models';


@Injectable({providedIn: 'root'})
export class PostsService {
  private baseUrl = 'http://localhost:3000/posts';

  constructor(private http: HttpClient) {}

  create(post: Post): Observable<Post> {
    return this.http.post<Post>(this.baseUrl, post);
  }

  update(post: Post): Observable<Post> {
    return this.http.put<Post>(`${this.baseUrl}/${post.uuid}`, post);
  }

  get(uuid: string): Observable<Post> {
    return this.http.get<Post>(`${this.baseUrl}/${uuid}`);
  }

  getAll(): Observable<Post[]> {
    return this.http.get<Post[]>(this.baseUrl);
  }

  delete(uuid: string): Observable<string> {
    return this.http.delete<string>(`${this.baseUrl}/${uuid}`).pipe(
      map(() => uuid)
    );
  }
}
