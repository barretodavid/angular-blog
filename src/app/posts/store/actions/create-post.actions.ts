import { Action } from '@ngrx/store';
import { Post } from '../../posts.models';

export enum CreatePostActionType {
  Start = '[Post] Create Post Start',
  Success = '[Post] Create Post Success',
  Error = '[Post] Create Post Error',
}

export class CreatePostStartAction implements Action {
  readonly type = CreatePostActionType.Start;
  constructor(public payload: Post) {}
}

export class CreatePostSuccessAction implements Action {
  readonly type = CreatePostActionType.Success;
  constructor(public payload: Post) {}
}

export class CreatePostErrorAction implements Action {
  readonly type = CreatePostActionType.Error;
}

export type CreatePostAction =
  | CreatePostStartAction
  | CreatePostSuccessAction
  | CreatePostErrorAction;
