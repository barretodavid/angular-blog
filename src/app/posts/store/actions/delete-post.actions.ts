import { Action } from '@ngrx/store';
import { Post } from '../../posts.models';

export enum DeletePostActionType {
  Start = '[Post] Delete Post Start',
  Success = '[Post] Delete Post Success',
  Error = '[Post] Delete Post Error',
}

export class DeletePostStartAction implements Action {
  readonly type = DeletePostActionType.Start;
  constructor(public payload: string) {}
}

export class DeletePostSuccessAction implements Action {
  readonly type = DeletePostActionType.Success;
  constructor(public payload: string) {}
}

export class DeletePostErrorAction implements Action {
  readonly type = DeletePostActionType.Error;
  constructor() {}
}

export type DeletePostAction = DeletePostStartAction | DeletePostSuccessAction | DeletePostErrorAction;
