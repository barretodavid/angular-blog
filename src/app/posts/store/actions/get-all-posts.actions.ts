import { Action } from '@ngrx/store';
import { Post } from '../../posts.models';

export enum GetAllPostsActionType {
  Start = '[Post] Get All Posts Start',
  Success = '[Post] Get All Posts Success',
  Error = '[Post] Get All Posts Error',
}

export class GetAllPostsStartAction implements Action {
  readonly type = GetAllPostsActionType.Start;
}

export class GetAllPostsSuccessAction implements Action {
  readonly type = GetAllPostsActionType.Success;
  constructor(public payload: Post[]) {}
}

export class GetAllPostsErrorAction implements Action {
  readonly type = GetAllPostsActionType.Error;
}

export type GetAllPostsAction =
  | GetAllPostsStartAction
  | GetAllPostsSuccessAction
  | GetAllPostsErrorAction;
