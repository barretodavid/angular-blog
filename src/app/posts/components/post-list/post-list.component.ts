import { Component, OnInit, ChangeDetectionStrategy } from '@angular/core';
import { PostSelector } from '../../store/posts.selectors';
import { Store } from '@ngrx/store';
import { State } from 'src/app/store/models';
import { GetAllPostsStartAction } from '../../store/actions/get-all-posts.actions';

@Component({
  selector: 'rg-post-list',
  template: `
    <mat-card class="mv2" *ngFor="let post of postSelector.posts$ | async">
      <mat-card-title>
        <a [routerLink]="['/edit-post', post.uuid]">{{ post.title }}</a>
      </mat-card-title>
      <mat-card-content>{{ post.content}}</mat-card-content>
    </mat-card>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class PostListComponent implements OnInit {

  constructor(
    public postSelector: PostSelector,
    private store: Store<State>
  ) {}

  ngOnInit() {
    this.postSelector.fetchAllPosts$.subscribe(() => {
      this.store.dispatch(new GetAllPostsStartAction());
    });
  }

}
