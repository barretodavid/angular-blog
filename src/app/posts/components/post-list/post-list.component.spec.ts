import { ComponentFixture, async, TestBed } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { CommonModule, APP_BASE_HREF } from '@angular/common';
import { MatCardModule } from '@angular/material';
import { RouterModule } from '@angular/router';

import { PostListComponent } from './post-list.component';
import { of, empty } from 'rxjs';
import { PostSelector } from '../../store/posts.selectors';
import { Store } from '@ngrx/store';
import { GetAllPostsStartAction } from '../../store/actions/get-all-posts.actions';
import { Post } from '../../posts.models';

function configureTestBed(postSelectorStub): Promise<any> {
  const storeStub = {
    dispatch: jest.fn(),
  };

  return TestBed.configureTestingModule({
    imports: [
      NoopAnimationsModule,
      CommonModule,
      MatCardModule,
      RouterModule.forRoot([]),
    ],
    providers: [
      { provide: APP_BASE_HREF, useValue: '/' },
      { provide: PostSelector, useValue: postSelectorStub },
      { provide: Store, useValue: storeStub },
    ],
    declarations: [PostListComponent],
  }).compileComponents();
}

describe('PostListComponent', () => {
  let postSelectorStub;

  describe('when there are no posts in the store', () => {
    beforeEach(() => {
      postSelectorStub = {
        fetchAllPosts$: of(null),
        posts$: of([]),
      };
    });

    test('should render an empty body', async () => {
      await configureTestBed(postSelectorStub);
      const fixture = TestBed.createComponent(PostListComponent);
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });

    test('should dispatch an action to get all posts', async () => {
      await configureTestBed(postSelectorStub);
      const fixture = TestBed.createComponent(PostListComponent);
      const storeStub = fixture.debugElement.injector.get(Store);
      fixture.detectChanges();
      expect(storeStub.dispatch).toHaveBeenCalledWith(
        new GetAllPostsStartAction(),
      );
    });
  });

  describe('when there is at least on blog post in the store', () => {
    beforeEach(() => {
      const post: Post = {
        uuid: '123',
        title: 'my title',
        content: 'my content',
      };

      postSelectorStub = {
        fetchAllPosts$: empty(),
        posts$: of([post]),
      };
    });

    test('it should render the post in the body', async () => {
      await configureTestBed(postSelectorStub);
      const fixture = TestBed.createComponent(PostListComponent);
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });

    test('it should not dispatch an action to load all post from server', async () => {
      await configureTestBed(postSelectorStub);
      const fixture = TestBed.createComponent(PostListComponent);
      const storeStub = fixture.debugElement.injector.get(Store);
      fixture.detectChanges();
      expect(storeStub.dispatch).not.toHaveBeenCalled();
    });
  });
});
