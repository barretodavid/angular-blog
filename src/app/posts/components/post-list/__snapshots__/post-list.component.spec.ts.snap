// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostListComponent when there are no posts in the store should render an empty body 1`] = `
<rg-post-list
  postSelector={[Function Object]}
  store={[Function Object]}
>
  
</rg-post-list>
`;

exports[`PostListComponent when there is at least on blog post in the store it should render the post in the body 1`] = `
<rg-post-list
  postSelector={[Function Object]}
  store={[Function Object]}
>
  <mat-card
    class="mv2 mat-card"
  >
    <mat-card-title
      class="mat-card-title"
    >
      <a
        href="/edit-post/123"
        ng-reflect-router-link="/edit-post,123"
      >
        my title
      </a>
    </mat-card-title>
    <mat-card-content
      class="mat-card-content"
    >
      my content
    </mat-card-content>
  </mat-card>
</rg-post-list>
`;
